Wikipedia/Code/WMFArticleListTableViewController.m

```diff
@@ -259,4 +252,83 @@ - (void)tableViewUpdater:(WMFTableViewUpdater *)updater didUpdateTableView:(UITa
     });
 }
 
+#pragma mark - Sharing
+
+- (UIActivityViewController *)shareActivityController:(NSURL *)url {
+    WMFArticle *article = [self.userDataStore fetchArticleWithURL:url];
+    NSMutableArray *items = [NSMutableArray array];
+    
+    
+    // TODO: Use WMFArticleTextActivitySource
+    NSString *text =  [NSString stringWithFormat:@"\"%@\" on @Wikipedia", [article displayTitle]];
+    [items addObject:text];
+    
+    NSURL *desktopURL = [NSURL wmf_desktopURLForURL:url];
+    if (desktopURL) {
+        NSURLComponents *components = [NSURLComponents componentsWithURL:url resolvingAgainstBaseURL:NO];
+        
+        NSURLQueryItem *queryItem = [NSURLQueryItem queryItemWithName:@"wprov" value:@"sfsi1"];
+        components.queryItems = @[queryItem];
+        
+        NSURL *componentsURL = components.URL;
+        if (componentsURL) {
+            [items addObject:componentsURL];
+        }
+    }
+    
+    MKMapItem *mapItem = [article mapItem];
+    if (mapItem) {
+        [items addObject:mapItem];
+    }
+
+    UIActivityViewController *vc = [[UIActivityViewController alloc] initWithActivityItems:items
+                                                                                              applicationActivities:
+                                                             @[[[TUSafariActivity alloc] init],
+                                                               [[WMFOpenInMapsActivity alloc] init],
+                                                               [[WMFGetDirectionsInMapsActivity alloc] init]]];
+    return vc;
+}
+
+- (void)shareArticle:(UIActivityViewController *)shareActivityController {
+    [self presentViewController:shareActivityController animated:YES completion:NULL];
+}
+
+#pragma mark - Row actions
+
+- (NSString *)deleteActionText {
+    return WMFLocalizedStringWithDefaultValue(@"article-delete", nil, nil, @"Delete", @"Text of the article list row action shown on swipe which deletes the article");
+}
+
+- (NSString *)shareActionText {
+    return WMFLocalizedStringWithDefaultValue(@"article-share", nil, nil, @"Share", @"Text of the article list row action shown on swipe which allows the user to choose the sharing option");
+}
+
+- (NSString *)saveActionText {
+    return WMFLocalizedStringWithDefaultValue(@"article-save", nil, nil, @"Save", @"Text of the article list row action shown on swipe which allows the user to save the article");
+}
+
+- (UITableViewRowAction *)shareAction:(NSIndexPath *)indexPath {
```

Comment:
These row actions can lead to some conflicting animations. The simplest workaround would be to add a helper method to this VC that doesn't apply the edit until the table view end editing animation is complete (similar to how places handles it):

```
- (UITableViewRowAction *)rowActionWithStyle:(UITableViewRowActionStyle)style title:(nullable NSString *)title handler:(void (^)(UITableViewRowAction *action, NSIndexPath *indexPath))handler {
    return [UITableViewRowAction rowActionWithStyle:style title:title handler:^(UITableViewRowAction *action, NSIndexPath *indexPath){
        [CATransaction begin];
        [CATransaction setCompletionBlock:^{
            if (handler) {
                handler(action, indexPath);
            }
        }];
        [self.tableView setEditing:NO animated:YES];
        [CATransaction commit];
    }];
}
```

before:
![before](https://user-images.githubusercontent.com/741327/27961232-12ad84aa-62fc-11e7-833a-f8322c92c863.gif)

after:
![after](https://user-images.githubusercontent.com/741327/27961240-17ecbc2e-62fc-11e7-905c-43a83154d33d.gif)

Also, `WMFTableViewUpdater.m` should have `UITableViewRowAnimationNone` for any `reloadRowsAtIndexPaths:`